File: E_Rudolf_and_k_Bridges.py
from heapq import *

for _ in range(int(input())):
    n, m, k, d = map(int, input().split())
    pre = [0 for _ in range(n)]
    for i in range(n):
        a = list(map(int, input().split()))
        q = []
        heappush(q, (1, 0))
        for j in range(1, m):
            while q and q[0][1] < j - d - 1:
                heappop(q)
            heappush(q, (q[0][0] + a[j] + 1, j))
        while q and q[0][1] != m - 1:
            heappop(q)
        pre[i] = q[0][0]
    cnt = sum(pre[:k])
    ans = cnt
    for i in range(k, n):
        cnt += pre[i] - pre[i - k]
        ans = min(ans, cnt)
    print(ans)

File: E_Rudolf_and_k_Bridges.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int long long

void solve() {
    int n, m, k, d;
    cin >> n >> m >> k >> d;
    vector<int> pre(n);
    for (int i = 0; i < n; ++i) {
        vector<int> a(m);
        for (int j = 0; j < m; ++j) {
            cin >> a[j];
        }
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> q;
        q.emplace(1, 0);
        for (int j = 1; j < m; ++j) {
            while (q.size() and q.top().ss < j - d - 1) {
                q.pop();
            }
            q.emplace(q.top().ff + a[j] + 1, j);
        }
        while (q.size() and q.top().ss != m - 1) {
            q.pop();
        }
        pre[i] = q.top().ff;
    }
    int cnt = accumulate(pre.begin(), pre.begin() + k, 0ll);
    int ans = cnt;
    for (int i = k; i < n; ++i) {
        cnt += pre[i] - pre[i - k];
        ans = min(ans, cnt);
    }
    cout << ans << endl;
}

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    cout << fixed << setprecision(15);
    int Test; cin >> Test; while (Test--)
    solve();
    return 0;
}
